src/number_theory/diophantine_approximation.lean

```diff
@@ -0,0 +1,363 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Geißer, Michael Stoll
+-/
+import tactic.basic
+import data.real.irrational
+import combinatorics.pigeonhole
+
+/-!
+# Diophantine Approximation
+
+This file gives proofs of various versions of **Dirichlet's approximation theorem**
+and its important consequence that when `ξ` is an irrational real number, then there are
+infinitely many rationals `x/y` (in lowest terms) such that `|ξ - x/y| < 1/y^2`.
+
+We also show the converse, i.e., that for rational `ξ` there are only finitely many
+such rational approximations.
+
+The proof (of the interesting direction) is based on the pigeonhole principle.
+
+## Main definitions
+
+We define the set `dioph_approx.rat_approx ξ` for a real number `ξ` to be the set
+of pairs `(x, y)` of coprime integers with `y` positive such that `|ξ - x/y| < 1/y^2`.
+This set is in natural bijection with `{q : ℚ | |ξ - q| < 1/q.denom^2}`
+(see `dioph_approx.rat_approx_equiv`).
+
+## Main statements
+
+The main results are
+* `dioph_approx.dirichlet_approx`, which is a version of Dirichlet's approximation
+  theorem and states that for all real `ξ` and natural `0 < n`, there are integers
+  `j` and `k` with `0 < k ≤ n` and `|ξ*k - j| ≤ 1/(n+1)`,
+* A variant `dioph_approx.dirichlet_approx'` of this in terms of rationals `q`
+  satisfying `|ξ - q| ≤ 1/((n+1)*q.denom)` and `q.denom ≤ n`,
+* `dioph_approx.rat_approx_infinite`, which states that for irrational `ξ`,
+  `{q : ℚ | |ξ - q| < 1/q.denom^2}` is infinite,
+* `dioph_approx.rat_approx_finite`, which states that `{q : ℚ | |a/b - q| < 1/q.denom^2}`
+  is finite for integers `a` and `b`,
+* `dioph_approx.rat_approx_infinite_iff_irrational`, which combines the two previous
+  statements to give an iff statement, and
+* `dioph_approx.rat_approx_infinite_iff_irrational'`, which is a version
+   in terms of `dioph_approx.rat_approx ξ`.
+
+## Implementation notes
+
+We use the namespace `dioph_approx`.
+
+## References
+
+<https://en.wikipedia.org/wiki/Dirichlet%27s_approximation_theorem>
+
+## Tags
+
+Diophantine approximation, Dirichlet's approximation theorem
+-/
+
+namespace dioph_approx
+
+/-!
+### Dirichlet's approximation theorem
+
+We show that for any real number `ξ` and positive natural `n`, there is a fraction `q`
+such that `q.denom ≤ n` and `|ξ - q| ≤ 1/((n+1)*q.denom)`.
+-/
+
+section pigeonhole
+
+open finset int
+
+/-- *Dirichlet's approximation theorem:*
+For any real number `ξ` and positive natural `n`, there are integers `j` and `k`,
+with `0 < k ≤ n` and `|ξ*k - j| ≤ 1/(n+1)`. -/
+lemma dirichlet_approx (ξ : ℝ) {n : ℕ} (n_pos : 0 < n) :
+  ∃ j k : ℤ, 0 < k ∧ k ≤ n ∧ |ξ * k - j| ≤ 1 / (n + 1) :=
```

Comment:
I'll have to see if using `round` helps with the proof.
The disadvantage of writing `↑k * ξ` instead of `ξ * k` is that one needs to make the cast explicit, which makes it look less nice. As it is easy to switch the order when desired, I'm not sure the change is worth while.

Reply:
In my experience the ability to use `nsmul_eq_mul` / `zsmul_eq_mul` is more valuable than the fact that the elaboration algorithm happens to allow us to omit the `↑`.

I'd be surprised if the proof using `round` is not easier. It would be less satisfactory but we can always add a corollary with the statement suggested I suggested, since it would follow trivially using https://leanprover-community.github.io/mathlib_docs/algebra/order/floor.html#round_le.

Reply:
- I played around a bit, and it did not seem that using `round` directly helps with the proof (on the contrary). So I decided to simply add the version you suggest (and prove it using the version with two integers, which is easy enough).
- After thinking a bit more, I have decided that it does make sense to write both statements in terms of `↑k * ξ`.

Reply:
> it did not seem that using round directly helps with the proof (on the contrary)

OK fair enough. I remark in passing that it can be valuable to isolate the difficulty in cases like this since it may simply be a gap in the API for `round` which will then turn into a lemma that anyone can use.

Reply:
In my experiments, it came down to missing statements like
```lean
example (x y : ℝ) : fract (x + y) = fract x + fract y ∨ fract (x + y) = fract x + fract y - 1
```
(or rather, a similar one for subtraction), which are not completely immediate from what is available (or I'm just not smart enough...).

Reply:
The hope is that there should be sufficient API for `round` to avoid you having to work in terms of `fract`.

Failing this, I would turn the `example` above (and its subtraction counterpart) into lemmas which I would add to `alegbra/order/floor.lean` so that they are available for everyone. I think our library grows best when we fill gaps rather than step around them.

Reply:
I think the proof via the existence of two integers is actually the most natural one here; this comes quite directly from applying the pigeonhole principle (which is, as far as I know, in fact the standard proof). I also think it makes sense to have both versions available.

Quite separately, it would certainly make sense to extend the API for `fract` by adding these lemmas, but my Lean bandwidth is limited (I have a day job as a math professor that requires some attention...), and so I prefer to leave this to somebody else for the time being.